### DISCUSSION TITLE
Fail to run Vitest test with an error: "Vitest caught 1 unhandled error during the test run"

### DISCUSSION THREAD

**Discussion by tal-rofe** [vitest-dev/vitest#3229] [Q&A]
I have a function which I try to test using Vitest.

This is my function (I test the exported one):

```ts
import os from 'node:os';

import { AI_COMMIT_IGNORED_FILES } from '../constants/ai-commit';
import { asyncExec } from './os';

/**
 * The function returns array of files paths that are in staged mode
 * @returns array of files paths in staged mode
 */
const getStagedFiles = async () => {
	const gitCommand = 'git diff --name-only --cached --relative .';
	const { stdout: filesOutput, stderr } = await asyncExec(gitCommand);

	if (stderr) {
		throw new Error(stderr);
	}

	const filesList = filesOutput.split(os.EOL).filter(Boolean);

	return filesList;
};


/**
 * The function returns the "git diff" command output for relevant staged files in the commit
 * @returns "git diff" output as a string
 */
export const getStagedFilesDiff = async () => {
	const stagedFiles = await getStagedFiles();
	const filteredFiles = stagedFiles.filter((file) => !AI_COMMIT_IGNORED_FILES.includes(file));

	const gitCommand = `git diff --staged -- ${filteredFiles.join(' ')}`;
	const { stdout: diffOutput, stderr } = await asyncExec(gitCommand);

	if (stderr) {
		throw new Error(stderr);
	}

	return diffOutput;
};
```

And this is the `asyncExec` function:
```ts
import { exec } from 'node:child_process';
import util from 'node:util';

export const asyncExec = util.promisify(exec);
```

I wrote the following test:
```ts
import { describe, it, expect, vi } from 'vitest';

import { asyncExec } from '@/utils/os';
import { getStagedFilesDiff } from '@/utils/git-info';

vi.mock('@/utils/os');

describe('[utils/git-info]', () => {
	describe('getStagedFilesDiff()', () => {
		it('should throw an error when "asyncExec" throws', () => {
			vi.mocked(asyncExec).mockRejectedValueOnce(undefined);

			expect(() => getStagedFilesDiff()).toThrowError();
		});
	});
});
```

But then I got the error: 
```

Vitest caught 1 unhandled error during the test run.
This might cause false positive tests. Resolve unhandled errors to make sure your tests are not affected.


This error originated in "tests/utils/git-info.spec.ts" test file. It doesn't mean the error was thrown inside the file itself, but while it was running.
⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
 ELIFECYCLE  Test failed. See above for more details.
```

I have followed this: https://vitest.dev/api/expect.html#tothrowerror

So I don't understand why doesn't it work..?

I tried the following also:

```ts
		it('should throw an error when "asyncExec" throws', async () => {
			vi.mocked(asyncExec).mockRejectedValueOnce(undefined);

			await expect(() => getStagedFilesDiff()).rejects.toThrowError(undefined);
		});
```

But got the error: `AssertionError: expected [Function] to throw an error`.

Then I wrapped the `getStagedFilesDiff` function with `try {} catch {}` as follows:
```ts
export const getStagedFilesDiff = async () => {
	try {
		const stagedFiles = await getStagedFiles();
		const filteredFiles = stagedFiles.filter((file) => !AI_COMMIT_IGNORED_FILES.includes(file));

		if (filteredFiles.length === 0) {
			return null;
		}

		const gitCommand = `git diff --staged -- ${filteredFiles.join(' ')}`;
		const { stdout: diffOutput, stderr } = await asyncExec(gitCommand);

		if (stderr) {
			throw new Error(`Failed to get "git diff" output of staged files with error:\n${stderr}`);
		}

		return diffOutput;
	} catch {
		throw new Error();
	}
};
```

Then the test completed successfully (when using the previous test code, not the original one - first one).

**Comment by sheremet-va** [comment]
> I have followed this: [vitest.dev/api/expect.html#tothrowerror](https://vitest.dev/api/expect.html#tothrowerror)

You did not follow this.

Your function is async. It needs `rejects`.

<img width="640" alt="Screenshot 2023-04-21 at 14 17 24" src="https://user-images.githubusercontent.com/16173870/233633262-43bc4514-f909-49b9-b38e-84d54d7a76e3.png">

**Reply by tal-rofe** [reply]
@sheremet-va 
My lord, I tried it as well:
```ts
		it('should throw an error when "asyncExec" throws', async () => {
			vi.mocked(asyncExec).mockRejectedValueOnce(undefined);

			await expect(() => getStagedFilesDiff()).rejects.toThrowError(undefined);
		});
```

the error is: 
```
AssertionError: expected [Function] to throw an error
```

**Reply by tal-rofe** [reply]
@sheremet-va 
Should I convert this one to an issue?

**Reply by sheremet-va** [reply]
You cannot convert this to an issue because there is no reproduction.

**Reply by tal-rofe** [reply]
@sheremet-va 
Ok, my lord. I will try to create a reproduction repository, and then, I will create an issue. :)

**Reply by shadowCruz** [reply]
i feel your pain tal. I am having similar problem :)

**Comment by camboYY** [comment]
Vitest caught 14 unhandled errors during the test run.
this happens when i run test coverage

**Reply by p0wl90** [reply]
Same here with Quasar, always different Errors in always different files. So it is impossible to find the actual problem (we run around 1k tests).

**Reply by roccomilicic** [reply]
Ever fix it?

**Comment by piecyk** [comment]
Is there any way to catch what throws this error? Pretty hard to pinpoint it in large codebase.

**Reply by sheremet-va** [reply]
> Is there any way to catch what throws this error?

The error is called "unhandled". It means that nothing _could_ catch it

**Reply by piecyk** [reply]
> Vitest caught 1 unhandled error during the test run.

Something did ;) but jokes aside... Can't really pin it, mostly getting it randomly in netlify preview build, after manually triggering it again then it's ok, no errors when running it locally.

**Reply by p0wl90** [reply]
Here are some more infos on resolving this:

* https://github.com/vitest-dev/vitest/issues/1692#issuecomment-1293941666
* https://github.com/vitest-dev/vitest/issues/1720#issuecomment-1194950162

I did not find time to investigate it in our large codebase myself but maybe it helps you. I'd be thankful if you cold keep us updated!

**Comment by unional** [comment]
I'm getting this randomly (but very frequently) when I use `maxWorkers` and Vitest Browser Mode together. Without `maxWorkers`, the tests passes all the time.

**Comment by SimpleCookie** [comment]
I realise it's an old post, but if someone else finds this place and has the same error you could have the same issue as I did.
Meaning: Wrong node version. 

My tests ran fine in terminal where I explicitly set the correct node version. But my system default was different, so when I ran tests through the vscode extension vitest I got the same error "Vitest caught 1 unhandled error during the test run".

Hopefully this might help someone in the future :) 

Background: 
In my case I tracked it down to a package in jsdom -> whatwg-url
So it's probably something there that's uncompatible with earlier versions of node

**Reply by roccomilicic** [reply]
What version of node and vitest were you using to remove the unhandeld error?

**Reply by SimpleCookie** [reply]
> What version of node and vitest were you using to remove the unhandeld error?

Sorry I don't remember. I'm guessing v20

**Comment by kelvinwop** [comment]
For me it was some kind of strange memory leak over time. Found it via 

logHeapUsage: true,

Ended up fixing via switching   'vmForks' to straight up 'forks'. tests take 3x as long now but at least they're totally isolated from each other and work now lol
